certora/specs/ERC20Invariants.spec

```diff
@@ -0,0 +1,89 @@
+}
+
+// Ghost copy of DelegationTokenStorage._delegatee for quantification.
+ghost mapping(address => address) ghost_delegatee {
+    init_state axiom forall address account. ghost_delegatee[account] == 0;
+}
+
+// Slot for DelegationTokenStorage._delegatee.
+hook Sload address delegatee (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1100)[KEY address account] {
+    require ghost_delegatee[account] == delegatee;
+}
+
+// Slot for DelegationTokenStorage._delegatee.
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1100)[KEY address account] address delegatee (address delegateeOld) {
+    // Update partial sums for x > to_mathint(account)
+    // Track delegation changes from the parameterized address.
+    if (delegateeOld == A && delegatee != A) {
+        havoc sumsOfVotes assuming
+            forall mathint x. sumsOfVotes@new[x] ==
+            sumsOfVotes@old[x] - (to_mathint(account) < x ? ghost_balances[account] : 0);
+    }
+    // Track delegation changes to the pramaeterized address.
+    else if (delegateeOld != A && delegatee == A) {
+        havoc sumsOfVotes assuming
+            forall mathint x. sumsOfVotes@new[x] ==
+                sumsOfVotes@old[x] + (to_mathint(account) < x ? ghost_balances[account] : 0);
+    }
+    // Update ghost copy of DelegationTokenStorage._delegatee.
+    ghost_delegatee[account] = delegatee;
+}
+
+// Ghost variable to hold the sum of balances.
+ghost mathint sumOfBalances {
+    init_state axiom sumOfBalances == 0;
+}
+
+// Ghost copy of ERC20Storage._balances for quantification.
+ghost mapping(address => uint256) ghost_balances {
+    init_state axiom forall address account. ghost_balances[account] == 0;
+}
+
+//Slot is ERC20Storage._balances slot.
+hook Sload uint256 balance (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address account] {
+    require ghost_balances[account] == balance;
+    // Safe require as accounts can't hold more tokens than the total supply in preconditions.
+    require sumOfBalances >= to_mathint(balance);
+}
+
+//Slot is ERC20Storage._balances slot
+hook Sstore (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address account] uint256 newValue (uint256 oldValue) {
+    // Update partial sums for x > to_mathint(account)
+    // Track balance changes when the delegatee is the parameterized address.
+    if (ghost_delegatee[account] == A) {
+        havoc sumsOfVotes assuming
+            forall mathint x. sumsOfVotes@new[x] ==
+                sumsOfVotes@old[x] + (to_mathint(account) < x ? newValue - oldValue : 0);
+    }
+    // Update ghost copy of ERC20Storage._balances.
+    ghost_balances[account] = newValue;
+    // Track changes of total supply.
+    sumOfBalances = sumOfBalances - oldValue + newValue;
+}
+
+// Check that the sum of balances equals the total supply.
+invariant totalSupplyIsSumOfBalances()
+    to_mathint(totalSupply()) == sumOfBalances;
```

Comment:
If we have only this then we did not gain anything compared to the previous invariant we had. Previously (before this PR), we already had `totalSupply() == sumOfBalances`. The difference is how is specified `sumOfBalances`, but I wouldn't say that it's simpler now. The benefit we get from this specification is to be able to prove that `sumOfBalances` is greater than the sum of n number of accounts balances

Reply:
I've added this proof in https://github.com/morpho-org/morpho-token-upgradeable/pull/102/commits/1c593dc2a9a833a770e3ff1373125b99467073de.